<?xml version="1.0" encoding="UTF-8"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="UMIUSI">

  <xacro:macro name="thruster_direct"
    params="
    esc1_pin esc2_pin esc3_pin esc4_pin
    servo1_pin servo2_pin servo3_pin servo4_pin
    esc1_center_pulse_width esc1_negative_pulse_width_radius esc1_positive_pulse_width_radius
    esc2_center_pulse_width esc2_negative_pulse_width_radius esc2_positive_pulse_width_radius
    esc3_center_pulse_width esc3_negative_pulse_width_radius esc3_positive_pulse_width_radius
    esc4_center_pulse_width esc4_negative_pulse_width_radius esc4_positive_pulse_width_radius
    servo1_min_pulse_width servo1_max_pulse_width
    servo2_min_pulse_width servo2_max_pulse_width
    servo3_min_pulse_width servo3_max_pulse_width
    servo4_min_pulse_width servo4_max_pulse_width">
    <xacro:macro name="thruster_direct"
      params="id esc_pin servo_pin
      esc_center_pulse_width esc_negative_pulse_width_radius esc_positive_pulse_width_radius
      servo_min_pulse_width servo_max_pulse_width">
      <xacro:property name="thruster_name" value="thruster_direct${id}" />
      <ros2_control name="${thruster_name}/servo" type="system">
        <hardware>
          <plugin>sinsei_umiusi_control/hardware/ServoDirect</plugin>
          <param name="id">${id}</param>
          <param name="pin">${servo_pin}</param>
          <param name="min_pulse_width">${servo_min_pulse_width}</param>
          <param name="max_pulse_width">${servo_max_pulse_width}</param>
        </hardware>
        <gpio name="${thruster_name}/servo">
          <command_interface name="enabled" />
          <command_interface name="angle" />
          <state_interface name="health" />
        </gpio>
      </ros2_control>
      <ros2_control name="${thruster_name}/esc" type="system">
        <hardware>
          <plugin>sinsei_umiusi_control/hardware/EscDirect</plugin>
          <param name="id">${id}</param>
          <param name="pin">${esc_pin}</param>
          <param name="center_pulse_width">${esc_center_pulse_width}</param>
          <param name="negative_pulse_width_radius">${esc_negative_pulse_width_radius}</param>
          <param name="positive_pulse_width_radius">${esc_positive_pulse_width_radius}</param>
        </hardware>
        <gpio name="${thruster_name}/esc">
          <command_interface name="enabled" />
          <command_interface name="duty_cycle" />
          <state_interface name="health" />
        </gpio>
      </ros2_control>
    </xacro:macro>

    <xacro:thruster_direct id="1" esc_pin="${esc1_pin}" servo_pin="${servo1_pin}"
      esc_center_pulse_width="${esc1_center_pulse_width}"
      esc_negative_pulse_width_radius="${esc1_negative_pulse_width_radius}"
      esc_positive_pulse_width_radius="${esc1_positive_pulse_width_radius}"
      servo_min_pulse_width="${servo1_min_pulse_width}"
      servo_max_pulse_width="${servo1_max_pulse_width}"
    />
    <xacro:thruster_direct id="2" esc_pin="${esc2_pin}" servo_pin="${servo2_pin}"
      esc_center_pulse_width="${esc2_center_pulse_width}"
      esc_negative_pulse_width_radius="${esc2_negative_pulse_width_radius}"
      esc_positive_pulse_width_radius="${esc2_positive_pulse_width_radius}"
      servo_min_pulse_width="${servo2_min_pulse_width}"
      servo_max_pulse_width="${servo2_max_pulse_width}"
    />
    <xacro:thruster_direct id="3" esc_pin="${esc3_pin}" servo_pin="${servo3_pin}"
      esc_center_pulse_width="${esc3_center_pulse_width}"
      esc_negative_pulse_width_radius="${esc3_negative_pulse_width_radius}"
      esc_positive_pulse_width_radius="${esc3_positive_pulse_width_radius}"
      servo_min_pulse_width="${servo3_min_pulse_width}"
      servo_max_pulse_width="${servo3_max_pulse_width}"
    />
    <xacro:thruster_direct id="4" esc_pin="${esc4_pin}" servo_pin="${servo4_pin}"
      esc_center_pulse_width="${esc4_center_pulse_width}"
      esc_negative_pulse_width_radius="${esc4_negative_pulse_width_radius}"
      esc_positive_pulse_width_radius="${esc4_positive_pulse_width_radius}"
      servo_min_pulse_width="${servo4_min_pulse_width}"
      servo_max_pulse_width="${servo4_max_pulse_width}"
    />

  </xacro:macro>

</robot>
